File: library/src/main/java/renetik/android/view/list/CSListLoadNextController.java
package renetik.android.view.list;

import android.widget.AbsListView;
import android.widget.AbsListView.OnScrollListener;
import android.widget.FrameLayout;
import android.widget.FrameLayout.LayoutParams;
import android.widget.ListView;

import java.util.List;

import renetik.android.viewbase.CSView;
import renetik.android.viewbase.CSViewController;
import renetik.android.java.event.CSEvent;

import static android.view.Gravity.BOTTOM;
import static android.view.Gravity.CENTER;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;
import static renetik.android.lang.CSLang.NO;
import static renetik.android.lang.CSLang.event;
import static renetik.android.lang.CSLang.fire;
import static renetik.android.lang.CSLang.no;

public class CSListLoadNextController extends CSViewController {

    public final CSEvent<Void> onLoadNext = event();
    private final CSView<?> _loadView;
    private EndlessScrollListener _scrollListener;
    private boolean _loading;

    public CSListLoadNextController(CSListController parent, int loadViewLayout) {
        super(parent);
        _loadView = new CSView(this, layout(loadViewLayout));
        parent.onLoad.add((registration, list) -> onListLoad((List) list));
    }

    private void onListLoad(List data) {
        _loadView.hide();
        if (no(_scrollListener)) _scrollListener = new EndlessScrollListener();
        else if (data.isEmpty()) _scrollListener = null;
        updateScrollListener();
        _loading = false;
    }

    private void updateScrollListener() {
        asAbsListView().setOnScrollListener(_scrollListener);
    }

    private void onLoadNext() {
        fire(onLoadNext);
        _loading = true;
        _loadView.show();
    }

    public void onResume() {
        super.onResume();
        updateScrollListener();
        if (getView() instanceof ListView) {
            asListView().addFooterView(_loadView.getView());
            asListView().setFooterDividersEnabled(NO);
        } else if (!_loadView.hasParent())
            ((FrameLayout) getView().getParent().getParent()).addView(_loadView.getView(),
                    new LayoutParams(WRAP_CONTENT, WRAP_CONTENT, BOTTOM | CENTER));
        _loadView.hide();
    }

    private class EndlessScrollListener implements OnScrollListener {
        private int _visibleThreshold = 3;

        public void onScrollStateChanged(AbsListView view, int scrollState) {
        }

        public void onScroll(AbsListView view, int first, int visible, int total) {
            if (total - visible <= 0 || _loading) return;
            if (total - visible <= first + _visibleThreshold) onLoadNext();
        }
    }
}

File: library/src/main/java/renetik/android/view/list/CSListLoadNextController.kt
package renetik.android.view.list

import android.view.Gravity.BOTTOM
import android.view.Gravity.CENTER
import android.view.View
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.widget.AbsListView
import android.widget.AbsListView.OnScrollListener
import android.widget.FrameLayout
import android.widget.FrameLayout.LayoutParams
import android.widget.ListView
import renetik.android.view.base.CSView
import renetik.android.view.base.CSViewController
import renetik.android.view.base.layout
import renetik.java.event.event
import renetik.java.event.fire
import renetik.java.extensions.isNull
import renetik.java.lang.CSLang

class CSListLoadNextController<ListType : AbsListView>(
        val parent: CSListController<*, ListType>, list: CSListController<*, ListType>, loadViewLayout: Int)
    : CSViewController<ListType>(list) {

    val onLoadNext = event<Unit>()
    private val loadView = CSView<View>(this, layout(loadViewLayout))
    private var scrollListener: EndlessScrollListener? = null
    private var loading = false

    init {
        parent.onLoad.run { _, list -> onListLoad(list) }
    }

    private fun onListLoad(data: List<*>) {
        loadView.hide()
        if (scrollListener.isNull) scrollListener = EndlessScrollListener()
        else if (data.isEmpty()) scrollListener = null
        updateScrollListener()
        loading = false
    }

    private fun updateScrollListener() = view.setOnScrollListener(scrollListener)

    private fun onLoadNext() {
        onLoadNext.fire()
        loading = true
        loadView.show()
    }

    public override fun onResume() {
        super.onResume()
        updateScrollListener()
        (view as? ListView)?.apply {
            addFooterView(loadView.view)
            setFooterDividersEnabled(CSLang.NO)
        } ?: let {
            if (!loadView.hasParent) (parent.view as FrameLayout).addView(loadView.view,
                    LayoutParams(WRAP_CONTENT, WRAP_CONTENT, BOTTOM or CENTER))
        }
        loadView.hide()
    }

    private inner class EndlessScrollListener : OnScrollListener {
        private val visibleThreshold = 3

        override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {}

        override fun onScroll(view: AbsListView, first: Int, visible: Int, total: Int) {
            if (total - visible <= 0 || loading) return
            if (total - visible <= first + visibleThreshold) onLoadNext()
        }
    }
}
